db/wal_edit.cc

```diff
@@ -0,0 +1,184 @@
+    switch (tag) {
+      case WalAdditionTag::kSize: {
+        uint64_t size = 0;
+        if (!GetVarint64(src, &size)) {
+          return Status::Corruption(class_name, "Error decoding WAL file size");
+        }
+        metadata_.SetSizeInBytes(size);
+        break;
+      }
+      // TODO: process future tags such as checksum.
+      case WalAdditionTag::kTerminate:
+        return Status::OK();
+      default: {
+        std::stringstream ss;
+        ss << "Unknown tag " << tag_value;
+        return Status::Corruption(class_name, ss.str());
+      }
+    }
+  }
+}
+
+JSONWriter& operator<<(JSONWriter& jw, const WalAddition& wal) {
+  jw << "LogNumber" << wal.GetLogNumber() << "SizeInBytes"
+     << wal.GetMetadata().GetSizeInBytes();
+  return jw;
+}
+
+std::ostream& operator<<(std::ostream& os, const WalAddition& wal) {
+  os << "log_number: " << wal.GetLogNumber()
+     << " size_in_bytes: " << wal.GetMetadata().GetSizeInBytes();
+  return os;
+}
+
+std::string WalAddition::DebugString() const {
+  std::ostringstream oss;
+  oss << *this;
+  return oss.str();
+}
+
+void WalDeletion::EncodeTo(std::string* dst) const {
+  PutVarint64(dst, number_);
+}
+
+Status WalDeletion::DecodeFrom(Slice* src) {
+  constexpr char class_name[] = "WalDeletion";
+
+  if (!GetVarint64(src, &number_)) {
+    return Status::Corruption(class_name, "Error decoding WAL log number");
+  }
+
+  return Status::OK();
+}
+
+JSONWriter& operator<<(JSONWriter& jw, const WalDeletion& wal) {
+  jw << "LogNumber" << wal.GetLogNumber();
+  return jw;
+}
+
+std::ostream& operator<<(std::ostream& os, const WalDeletion& wal) {
+  os << "log_number: " << wal.GetLogNumber();
+  return os;
+}
+
+std::string WalDeletion::DebugString() const {
+  std::ostringstream oss;
+  oss << *this;
+  return oss.str();
+}
+
+Status WalSet::AddWal(const WalAddition& wal) {
+  if (wal.GetMetadata().HasSize()) {
+    // The WAL must exist without size.
+    if (wals_.find(wal.GetLogNumber()) == wals_.end()) {
+      std::stringstream ss;
+      ss << "WAL " << wal.GetLogNumber() << " is not created before closing";
+      return Status::Corruption("WalSet", ss.str());
+    }
+    if (wals_[wal.GetLogNumber()].HasSize()) {
+      std::stringstream ss;
+      ss << "WAL " << wal.GetLogNumber() << " is closed more than once";
+      return Status::Corruption("WalSet", ss.str());
+    }
+  } else {
+    // The WAL must not exist beforehand.
+    if (wals_.find(wal.GetLogNumber()) != wals_.end()) {
+      std::stringstream ss;
+      ss << "WAL " << wal.GetLogNumber() << " is created more than once";
+      return Status::Corruption("WalSet", ss.str());
+    }
+  }
+  wals_[wal.GetLogNumber()] = wal.GetMetadata();
+  return Status::OK();
+}
+
+Status WalSet::AddWals(const WalAdditions& wals) {
+  Status s;
+  for (const WalAddition& wal : wals) {
+    s = AddWal(wal);
+    if (!s.ok()) {
+      break;
+    }
+  }
+  return s;
+}
+
+Status WalSet::DeleteWal(const WalDeletion& wal) {
```

Comment:
This could also be optimized by storing the iterator to eliminate repeated lookups for the same key.

Reply:
changed to unordered_map